### PR TITLE
DAOS-11624 test: Update server storage yaml configuration

### PR DESCRIPTION
**Pull request by phender**
In preparation for metadata on SSDs update the existing functional tests from using the legacy storage configuration specification to using the storage tier specification.

Skip-unit-tests: true
Test-tag: aggregatebasic aggregationchecksum test_dfusespacecheck aggregateiosmall aggregate_single_pool

Required-githooks: true

Signed-off-by: Phil Henderson <phillip.henderson@intel.com>

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by phender** [comment]
Issues seen in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/8:
- : https://daosio.atlassian.net/browse/DAOS-11907
- : https://daosio.atlassian.net/browse/DAOS-11909
- : https://daosio.atlassian.net/browse/DAOS-11441 / https://daosio.atlassian.net/browse/DAOS-11376
- harness/advanced.py: (not normally run tests; timeouts) https://daosio.atlassian.net/browse/DAOS-11306
- pool/svc.py: https://daosio.atlassian.net/browse/DAOS-11724
- rebuild/cascading_failures.py: `No objects written to rank 3` - possibly a new error; will verify with next weekly results
- rebuild/container_rf.py: `No objects written to rank 3`
- server/daos_server_dump.py: manual tests that are expected to fail

**Comment by phender** [comment]
Analysis of https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/9 test failures:

- Errors from tests that intentionally fail or are not intended to run in CI (behaving as expected):
  - Functional on EL 8 / FTEST_launch.harness-advanced.103-./harness/advanced.py
  - Functional on EL 8 / no_cmocka_xml_file_test.4-./harness/basic.py:HarnessBasicTest.test_no_cmocka_xml
  - Functional Hardware Small / FTEST_launch.harness-advanced.28-./harness/advanced.py
  - Functional Hardware Medium / FTEST_launch.harness-advanced.30-./harness/advanced.py
  - Functional Hardware Large / FTEST_launch.harness-advanced.16-./harness/advanced.py
  - Functional Hardware Large / FTEST_performance.IorHard.2-./performance/ior_hard.py:IorHard.test_performance_ior_hard_dfs_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.IorHard.4-./performance/ior_hard.py:IorHard.test_performance_ior_hard_dfuse_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.MdtestEasy.2-./performance/mdtest_easy.py:MdtestEasy.test_performance_mdtest_easy_dfs_ec_16p2g1
  - Functional Hardware Large / FTEST_performance.MdtestEasy.5-./performance/mdtest_easy.py:MdtestEasy.test_performance_mdtest_easy_dfs_ec_16p2g1_stop
  - Functional Hardware Large / FTEST_performance.MdtestHard.2-./performance/mdtest_hard.py:MdtestHard.test_performance_mdtest_hard_dfs_ec_16p2g1

- https://daosio.atlassian.net/browse/DAOS-11925
  - Functional on EL 8 / FTEST_launch.datamover-dst_create.12-./datamover/dst_create.py
  - Functional Hardware Large / FTEST_launch.datamover-obj_large_posix.04-./datamover/obj_large_posix.py

- https://daosio.atlassian.net/browse/DAOS-11907
  - Functional on EL 8 / FTEST_container.AutoOCSelectionTest.1-./container/auto_oc_selection.py:AutoOCSelectionTest.test_oc_selection

- https://daosio.atlassian.net/browse/DAOS-11909
  - Functional on EL 8 / FTEST_datamover.DmvrDstCreate.1-./datamover/dst_create.py:DmvrDstCreate.test_dm_dst_create_dcp_posix_dfs
  - Functional on EL 8 / FTEST_datamover.DmvrDstCreate.2-./datamover/dst_create.py:DmvrDstCreate.test_dm_dst_create_dcp_posix_daos
  - Functional Hardware Large / FTEST_datamover.DmvrObjLargePosix.1-./datamover/obj_large_posix.py:DmvrObjLargePosix.test_dm_obj_large_posix_dcp

- https://daosio.atlassian.net/browse/DAOS-11928
  - Functional on EL 8 / FTEST_rebuild.RbldContRfTest.3-./rebuild/container_rf.py:RbldContRfTest.test_rebuild_with_container_rf
  - Functional on EL 8 / FTEST_rebuild.RbldContRfTest.4-./rebuild/container_rf.py:RbldContRfTest.test_rebuild_with_container_rf

- https://daosio.atlassian.net/browse/DAOS-11446
  - Functional Hardware Large / FTEST_erasurecode.EcodFioRebuild.07-./erasurecode/rebuild_fio.py:EcodFioRebuild.test_ec_online_rebuild_fio

- https://daosio.atlassian.net/browse/DAOS-11651
  - Functional Hardware Large / FTEST_erasurecode.EcodFioRebuild.11-./erasurecode/rebuild_fio.py:EcodFioRebuild.test_ec_online_rebuild_fio

- https://daosio.atlassian.net/browse/DAOS-11780
  - Functional Hardware Large / FTEST_ior.EcodIorHardRebuild.2-./ior/hard_rebuild.py:EcodIorHardRebuild.test_ec_ior_hard_online_rebuild

- https://daosio.atlassian.net/browse/DAOS-11927
  - Functional Hardware Large / FTEST_nvme.NvmePoolExclude.1-./nvme/pool_exclude.py:NvmePoolExclude.test_nvme_pool_excluded
  - Functional Hardware Large / FTEST_pool.PoolRedunFacProperty.1-./pool/rf.py:PoolRedunFacProperty.test_rf_pool_property

- https://daosio.atlassian.net/browse/DAOS-10769
  - Functional Hardware Large / FTEST_rebuild.RbldWidelyStriped.1-./rebuild/widely_striped.py:RbldWidelyStriped.test_rebuild_widely_striped

The following errors need to be fixed in ths PR:
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.1-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.2-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.3-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.4-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.5-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.6-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Small / FTEST_control.ConfigGenerateRun.7-./control/config_generate_run.py:ConfigGenerateRun.test_config_generate_run
  - Functional Hardware Large / FTEST_deployment.AgentFailure.1-./deployment/agent_failure.py:AgentFailure.test_agent_failure
  - Functional Hardware Large / FTEST_deployment.AgentFailure.2-./deployment/agent_failure.py:AgentFailure.test_agent_failure_isolation
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.01-./pool/create_all_hw.py:PoolCreateAllHwTests.test_one_pool
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.02-./pool/create_all_hw.py:PoolCreateAllHwTests.test_one_pool
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.03-./pool/create_all_hw.py:PoolCreateAllHwTests.test_one_pool
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.04-./pool/create_all_hw.py:PoolCreateAllHwTests.test_one_pool
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.05-./pool/create_all_hw.py:PoolCreateAllHwTests.test_recycle_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.06-./pool/create_all_hw.py:PoolCreateAllHwTests.test_recycle_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.07-./pool/create_all_hw.py:PoolCreateAllHwTests.test_recycle_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.08-./pool/create_all_hw.py:PoolCreateAllHwTests.test_recycle_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.09-./pool/create_all_hw.py:PoolCreateAllHwTests.test_two_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.10-./pool/create_all_hw.py:PoolCreateAllHwTests.test_two_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.11-./pool/create_all_hw.py:PoolCreateAllHwTests.test_two_pools
  - Functional Hardware Large / FTEST_pool.PoolCreateAllHwTests.12-./pool/create_all_hw.py:PoolCreateAllHwTests.test_two_pools
  - Functional Hardware Large / FTEST_performance.IorEasy.01-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_sx
  - Functional Hardware Large / FTEST_performance.IorEasy.02-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.IorEasy.03-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfuse_sx
  - Functional Hardware Large / FTEST_performance.IorEasy.04-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfuse_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.IorEasy.05-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_4p2gx_stop_write
  - Functional Hardware Large / FTEST_performance.IorEasy.06-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_4p2gx_stop_read
  - Functional Hardware Large / FTEST_performance.IorEasy.07-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx_stop_write
  - Functional Hardware Large / FTEST_performance.IorEasy.08-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx_stop_read
  - Functional Hardware Large / FTEST_performance.IorEasy.09-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_hdf5_sx
  - Functional Hardware Large / FTEST_performance.IorEasy.10-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_mpiio_sx

**Comment by phender** [comment]
Testing of updated files from build 9 passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/10/.  Build 10 did have the following expected errors:

- Errors from tests that fail due to requiring more resources than supported by CI (18 engines):
  - Functional Hardware Large / FTEST_performance.IorEasy.02-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.IorEasy.04-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfuse_ec_16p2gx
  - Functional Hardware Large / FTEST_performance.IorEasy.07-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx_stop_write
  - Functional Hardware Large / FTEST_performance.IorEasy.08-./performance/ior_easy.py:IorEasy.test_performance_ior_easy_dfs_ec_16p2gx_stop_read

**Comment by phender** [comment]
Summary of testing:
- https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/9
  - ran `all` tests
  - of the 58 failures, all but 31, where known issues (see previous comments for details)
- https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/10/
  - ran the 31 failed tests from build 9
  - all but 4 passed
  - the 4 failures are expected as they require more resources than are provided by CI (see previous comments for details)
- the only other failure is a CodeSpell error for a file not modified by this PR

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/util/server_utils_params.py [hunk 1]
```diff
@@ -319,202 +314,455 @@ def update_log_files(self, control_log, helper_log, server_log):
...
+        update = False
+        env_var_dict = {env.split("=")[0]: env.split("=")[1] for env in self.env_vars.value}
+        for key in sorted(required_env_vars):
+            if key not in env_var_dict or env_var_dict[key] != required_env_vars[key]:
+                env_var_dict[key] = required_env_vars[key]
+                update = True
+        if update:
+            self.log.debug("Assigning required env_vars")
+            new_env_vars = ["=".join([key, str(value)]) for key, value in env_var_dict.items()]
+            self.env_vars.update(new_env_vars, "env_var")
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils_params.py:439]: Would `EnvironmentVariables.from_list` or `EnvironmentVariables.update_from_list` be helpful here?

File: src/tests/ftest/util/server_utils_params.py [hunk 2]
```diff
@@ -319,202 +314,455 @@ def update_log_files(self, control_log, helper_log, server_log):
...
+
+    @property
+    def using_dcpm(self):
+        """Is the configuration file setup to use SCM devices.
+
+        Returns:
+            bool: True if SCM devices are configured; False otherwise
+
+        """
+        return self.storage.using_dcpm
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils_params.py:459]: Are these just convenience functions? For the `storage.*` equivalents?

File: src/tests/ftest/aggregation/checksum.yaml [hunk 3]
```diff
@@ -4,12 +4,18 @@ hosts:
...
-    bdev_class: nvme
-    bdev_list: ["0000:81:00.0"]
+  engines_per_host: 1
+  engines:
+    0:
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos
+          scm_size: 16
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/aggregation/checksum.yaml:14]: Yes, there is no longer a default for `class`, `scm_mount`, or `scm_size`.

File: src/tests/ftest/container/query_attribute.yaml [hunk 4]
```diff
@@ -3,10 +3,16 @@ hosts:
...
 server_config:
   name: daos_server
-  servers:
-    targets: 8
-    nr_xs_helpers: 0
-    scm_size: 4
+  engines_per_host: 1
+  engines:
+    0:
+      targets: 4
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/container/query_attribute.yaml:9]: No, but it didn't fail: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-10555/9/artifact/Functional%20on%20EL%208/container/query_attribute.py/results.html

File: src/tests/ftest/pool/create_all_hw.py [hunk 5]
```diff
@@ -29,7 +29,7 @@ def setUp(self):
 
         self.scm_avail_bytes, self.smd_avail_bytes = self.get_pool_available_bytes()
 
-        self.engine_target_size = int(self.params.get("targets", "server_config/servers/0/*"))
+        self.engine_target_size = int(self.server_managers[-1].get_config_value("targets"))
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/pool/create_all_hw.py:32]: There is only one server group defined for this test, so `self.server_managers[0] ==  self.server_managers[-1]`

File: src/tests/ftest/server/storage_tiers.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a DaosServerConfigTest object."""
+        super().__init__(*args, **kwargs)
+        self.start_agents_once = False
+        self.start_servers_once = False
+        self.setup_start_agents = False
+        self.setup_start_servers = False
+
+    def test_tiers(self):
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/storage_tiers.py:31]: The issue I have with including the ftest directory name in the test method name is it makes the avocado test name longer and repetitive:
```
Currently:       1-./server/storage_tiers.py:StorageTiers.test_tiers
With the change: 1-./server/storage_tiers.py:StorageTiers.test_server_storage_tiers
```

File: src/tests/ftest/server/storage_tiers.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            Verify storage tiers are correctly obtained from the test yaml file.
+
+        :avocado: tags=all,pr
+        :avocado: tags=hw,small
+        :avocado: tags=server,storage,storage_tiers
+        :avocado: tags=test_tiers
+        """
+        expected = [[], []]
+        for engine in range(2):
+            for tier in range(3):
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/storage_tiers.py:44]: Indeed, there are.

File: src/tests/ftest/server/storage_tiers.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                    if data["engines"][engine]["storage"] != expected[engine][0:max_tiers]:
+                        errors.append(f"{description}s do not match")
+                        self.log.error("  %s", errors[-1])
+                        errors.append(f"  actual:   {data['engines'][engine]['storage']}")
+                        self.log.error("  %s", errors[-1])
+                        errors.append(f"  expected: {expected[engine][0:max_tiers]}")
+                        self.log.error("  %s", errors[-1])
+                        continue
+                self.log.info("  %s passed", description)
+        if errors:
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/storage_tiers.py:88]: This is the pass message for the storage per engine.  The pass message for the entire test is on line 94.

[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/storage_tiers.py:88]: This is the pass message for the storage per engine loop.  If an error is detected then either `continue` will skip displaying this message.  The overall test pass message is displayed on line 94.

[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/storage_tiers.py:88]: Either `continue` on line 77 or line 86 should prevent this message from being logged in the case of a detected error.

File: src/tests/ftest/util/server_utils_params.py (hunk 2 above)
Comment (phender) [reply src/tests/ftest/util/server_utils_params.py:459]: Yes and they also existed for the previous `PerEngineYamlParameters` class.

File: src/tests/ftest/util/server_utils_params.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/util/server_utils_params.py:439]: Possibly, but with everything else going on in this PR I wasn't looking to change this code.

File: src/tests/ftest/server/storage_tiers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/tests/ftest/server/storage_tiers.py:31]: We should probably have a discussion at some point about what we want our standard to be

**Review by daltonbohning — APPROVED** [review]

**Comment by jolivier23** [comment]
One question, should this be on the branch?

**Pull request merged**